app/src/main/java/com/keylesspalace/tusky/util/LinkHelper.kt

```diff
@@ -70,27 +71,113 @@ fun getDomain(urlString: String?): String {
  * @param content containing text with mentions, links, or hashtags
  * @param mentions any '@' mentions which are known to be in the content
  * @param listener to notify about particular spans that are clicked
+ * @param trailingHashtagView a text view to fill with trailing / out-of-band hashtags
  */
 fun setClickableText(
     view: TextView,
     content: CharSequence,
     mentions: List<Mention>,
     tags: List<HashTag>?,
-    listener: LinkListener
+    listener: LinkListener,
+    trailingHashtagView: TextView?,
 ) {
     val spannableContent = markupHiddenUrls(view, content)
+    val (endOfContent, trailingHashtags) = when {
+        trailingHashtagView == null || tags.isNullOrEmpty() -> Pair(spannableContent.length, emptyList())
+        else -> getTrailingHashtags(spannableContent)
+    }
+    var inlineHashtagSpanCount = 0
 
     view.text = spannableContent.apply {
         styleQuoteSpans(view)
-        getSpans(0, spannableContent.length, URLSpan::class.java).forEach { span ->
+        getSpans(0, endOfContent, URLSpan::class.java).forEach { span ->
+            if (get(getSpanStart(span)) == '#') {
+                inlineHashtagSpanCount += 1
+            }
             setClickableText(span, this, mentions, tags, listener)
         }
-    }
+    }.subSequence(0, endOfContent).trimEnd()
+
     view.movementMethod = NoTrailingSpaceLinkMovementMethod
+
+    val showHashtagBar = (trailingHashtags.isNotEmpty() || inlineHashtagSpanCount != tags?.size)
+    // I don't _love_ setting the visibility here, but the alternative is to duplicate the logic in other places
+    trailingHashtagView?.visible(showHashtagBar)
+
+    if (showHashtagBar) {
+        trailingHashtagView?.apply {
+            text = SpannableStringBuilder().apply {
+                var offset = 0
+                tags?.forEachIndexed { index, tag ->
+                    val text = "#${tag.name}"
+                    append(text)
+                    setSpan(
+                        getCustomSpanForTag(text, tags, URLSpan(tag.url), listener),
+                        offset,
+                        offset + tag.name.length + 1,
+                        0
+                    )
+                    offset += tag.name.length + 1
```

Comment:
`SpannableStringBuilder` has a method [`append` that takes the text and the span at once](https://developer.android.com/reference/android/text/SpannableStringBuilder#append(java.lang.CharSequence,%20java.lang.Object,%20int)), using it should get rid of the offset calculations here